feat: Bokeh support as FigureBokeh

rileythai opened the pull request:
### All Submissions:
* [X] docs/examples
* [X] pre-commit
* [ ] tests
### TODO
* [X] fix printing issue
* [X] document common issues and their relevant workarounds
### Description of changes

This PR resolves #202, and adds a `solara.lab` component to support Bokeh `figure` and `Plot` objects. It additionally adds dependencies for `bokeh` and `jupyter_bokeh`. 

Part of the code is based on the implementation written in #202 by @maartenbreddels.

It behaves similarly to `FigurePlotly`. It takes arguments of `light_theme` and `dark_theme` to update the theme via a solara-side `effect` (the document instance must exist for us to update theming via the high-level interface)

This PR adds:
```
├── solara
│  └── lab
│     └── components
│        ├── bokehloaded.vue
│        └── figurebokeh.py
└───── website
      ├── pages
      │  ├── apps
      │  └────  scatter-bokeh.py
      └──── documentation
            └── examples
               ├── fullscreen
               │  └── scatter_bokeh.py
               └── visualization
                  └── bokeh.py
```

and modifies relevant files for importing components in `lab`

### Known issues/pitfalls
- Bokeh does not support an axis type change, so changing axes via high-level interface `x_axis_type`, etc doesnt work.
  - log workaround: update `x_scale` and/or `y_scale` of figure via a `use_effect`
  - categorical workaround: scale/map data server-side + update tickers directly, see the scatter-bokeh.py example
 - JS-side selection events on a `Tool` will fail to link to attributes of any `bokeh.model` if it is regenerated with each render.
   - occurs `bokeh.model.tools` instance with a callback like `CustomJS(args=dict(object=object))`, such as `object.selected.indices` or similar from a `ColumnDataSource` upon a selection.
    - suggested approach: `use_memo` models with callbacks to exclude from regeneration, then use `use_effect` or similar for attribute updates.

rileythai committed:
feat: bokeh support

maartenbreddels commented on solara/website/pages/documentation/examples/visualization/bokeh.py:
```diff
@@ -0,0 +1,48 @@
+def Page():
+    all_species = df["species"].unique().tolist()
+    species = solara.use_reactive(all_species[0])
+    with solara.Div() as main:
```
```suggestion
    with solara.Div():
```

maartenbreddels commented on solara/website/pages/documentation/examples/visualization/bokeh.py:
```diff
@@ -0,0 +1,48 @@
+        p.add_layout(cb, "right")
+
+        solara.lab.FigureBokeh(p, dark_theme="carbon", dependencies=[species])
+    return main
```
```suggestion
```

No need to use main/return main

maartenbreddels commented on solara/lab/components/figurebokeh.py:
```diff
@@ -0,0 +1,73 @@
+    light_theme: str | Theme = "light_minimal",
+    dark_theme: str | Theme = "dark_minimal",
+):
+    # NOTE: no docstring because not a component.
```
Why is this not a component? Because if we use this function conditionally, we'll have conditional hooks

rileythai replied:
It's the same reason `FigurePlotly` isn't a component. If you want to attach an effect callback to the figure widget, you won't be able to access it with `sl.get_widget` outside of the function itself.

If you make it a component, you'll get an error on `sl.get_widget` because it searches for the `FigureBokeh` and  not the `BokehModel` instance  
```python
 raise KeyError(f"Element {el} not found in all known widgets")  # for the component {context.widgets}")
KeyError: "Element FigureBokeh(Plot(id='p2331', ...), ...) not found in all known widgets"
```

rileythai replied:
I end up running into issues with https://github.com/widgetti/solara/issues/927 when using an effect outside if it is a component.

rileythai replied:
I ran into issues relating to #927 on occassion, couldn't track down what caused it

maartenbreddels reviewed (commented):
Nice! I think @Jhsmit will also like this. Maybe @Jhsmit can also review this?

rileythai committed:
fix: no main return in example

Co-authored-by: Maarten Breddels <maartenbreddels@gmail.com>

rileythai committed:
fix: no main return in example ((part II)

Co-authored-by: Maarten Breddels <maartenbreddels@gmail.com>

rileythai committed:
fix: FigureBokeh as component

- figurebokeh is now a component

rileythai committed:
feat: FigureBokeh docs page

- now a page in lab

rileythai committed:
fix: conditional hook (provide) in scatter-bokeh example

rileythai committed:
fix: routing for scatter-bokeh example

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

rileythai committed:
fix: patched BokehModel

- allow to track main

rileythai committed:
Merge branch 'master' into figurebokeh

rileythai committed:
fix: match arguments

- they have underscores

rileythai committed:
doc: add pitfalls to FigureBokeh

- these are inherent limitations (i think) to how bokeh is implemented
and how we have to work around that
